src/main/kotlin/org/codefreak/codefreak/service/file/JpaFileService.kt

```diff
@@ -48,4 +56,224 @@ class JpaFileService : FileService {
   override fun getCollectionMd5Digest(collectionId: UUID): ByteArray {
     return DigestUtils.md5Digest(getCollection(collectionId).tar)
   }
+
+  override fun createFile(collectionId: UUID, path: String) {
+    val normalizedPath = TarUtil.normalizeEntryName(path).withoutTrailingSlash()
+
+    requireValidPattern(path)
+    requireFileDoesNotExist(collectionId, normalizedPath)
+
+    getTarOutputStream(collectionId).use {
+      val input = getTarInputStream(collectionId)
+      TarUtil.copyEntries(input, it)
+      TarUtil.touch(normalizedPath, it)
+    }
+  }
+
+  private fun requireValidPattern(path: String) {
+    if (path.isBlank()) {
+      throw IllegalArgumentException("$path is not a valid path pattern")
+    }
+  }
+
+  private fun requireFileDoesExist(collectionId: UUID, path: String) {
+    if (!containsFile(collectionId, path)) {
+      throw IllegalArgumentException("$path does not exist or is no file")
+    }
+  }
+
+  private fun requireFileDoesNotExist(collectionId: UUID, path: String) {
+    if (containsFile(collectionId, path)) {
+      throw IllegalArgumentException("$path already exists")
+    }
+  }
+
+  private fun requireDirectoryDoesExist(collectionId: UUID, path: String) {
+    if (!containsDirectory(collectionId, path)) {
+      throw IllegalArgumentException("$path does not exist or is no directory")
+    }
+  }
+
+  private fun requireDirectoryDoesNotExist(collectionId: UUID, path: String) {
+    if (containsDirectory(collectionId, path)) {
+      throw IllegalArgumentException("$path already exists")
+    }
+  }
+
+  override fun containsFile(collectionId: UUID, path: String): Boolean {
+    val normalizedPath = TarUtil.normalizeEntryName(path).withoutTrailingSlash()
+    return containsPath(collectionId, normalizedPath) { it.isFile }
+  }
+
+  override fun containsDirectory(collectionId: UUID, path: String): Boolean {
+    val normalizedPath = TarUtil.normalizeEntryName(path).withTrailingSlash()
+    return containsPath(collectionId, normalizedPath) { it.isDirectory }
+  }
+
+  private fun containsPath(
+    collectionId: UUID,
+    path: String,
+    restriction: (TarArchiveEntry) -> Boolean = { true }
+  ): Boolean = findEntry(collectionId, path, restriction) != null
+
+  private fun findEntry(
+    collectionId: UUID,
+    path: String,
+    restriction: (TarArchiveEntry) -> Boolean = { true }
+  ): TarArchiveEntry? {
+    getTarInputStream(collectionId).use {
+      do {
```

Comment:
Koltin has a neat feature called Sequences that can be generated by such `next` methods:
https://kotlinlang.org/docs/reference/sequences.html#from-function
This allows more easy mapping and filtering

I think it is used already in the `TarUtil`